### 1. Import necesary libraries

In [1]:
from bs4 import BeautifulSoup
import requests
import pandas as pd

### 2. Get the HTML for the website

In [2]:
request = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text

### 3. Scrape the website using BeautifulSoup

In [3]:
soup = BeautifulSoup(request, 'lxml')
print(soup.prettify())

<!DOCTYPE html>
<html class="client-nojs" dir="ltr" lang="en">
 <head>
  <meta charset="utf-8"/>
  <title>
   List of postal codes of Canada: M - Wikipedia
  </title>
  <script>
   document.documentElement.className = document.documentElement.className.replace( /(^|\s)client-nojs(\s|$)/, "$1client-js$2" );
  </script>
  <script>
   (window.RLQ=window.RLQ||[]).push(function(){mw.config.set({"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":false,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":867606113,"wgRevisionId":867606113,"wgArticleId":539066,"wgIsArticle":true,"wgIsRedirect":false,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Communications in Ontario","Postal codes in Canada","Toronto","Ontario-related lists"],"wgBreakFrames":false,"wgPageContentLanguage":"en","wgPageContentModel":"wikitext","wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wg

### 4. Since our soup is cooked it's time fetch the table

In [4]:
table = soup.find('table')
print(table.prettify())

<table class="wikitable sortable">
 <tbody>
  <tr>
   <th>
    Postcode
   </th>
   <th>
    Borough
   </th>
   <th>
    Neighbourhood
   </th>
  </tr>
  <tr>
   <td>
    M1A
   </td>
   <td>
    Not assigned
   </td>
   <td>
    Not assigned
   </td>
  </tr>
  <tr>
   <td>
    M2A
   </td>
   <td>
    Not assigned
   </td>
   <td>
    Not assigned
   </td>
  </tr>
  <tr>
   <td>
    M3A
   </td>
   <td>
    <a href="/wiki/North_York" title="North York">
     North York
    </a>
   </td>
   <td>
    <a href="/wiki/Parkwoods" title="Parkwoods">
     Parkwoods
    </a>
   </td>
  </tr>
  <tr>
   <td>
    M4A
   </td>
   <td>
    <a href="/wiki/North_York" title="North York">
     North York
    </a>
   </td>
   <td>
    <a href="/wiki/Victoria_Village" title="Victoria Village">
     Victoria Village
    </a>
   </td>
  </tr>
  <tr>
   <td>
    M5A
   </td>
   <td>
    <a href="/wiki/Downtown_Toronto" title="Downtown Toronto">
     Downtown Toronto
    </a>
   </td>
   <td>
    <a href="

### 5.  Now that we found the table from our HTML we can now print the data from the table

In [5]:
info = []
for row in table.find_all('tr'):
    for cell in row.find_all('td'):
        info.append(cell.text)

In [6]:
info

['M1A',
 'Not assigned',
 'Not assigned\n',
 'M2A',
 'Not assigned',
 'Not assigned\n',
 'M3A',
 'North York',
 'Parkwoods\n',
 'M4A',
 'North York',
 'Victoria Village\n',
 'M5A',
 'Downtown Toronto',
 'Harbourfront\n',
 'M5A',
 'Downtown Toronto',
 'Regent Park\n',
 'M6A',
 'North York',
 'Lawrence Heights\n',
 'M6A',
 'North York',
 'Lawrence Manor\n',
 'M7A',
 "Queen's Park",
 'Not assigned\n',
 'M8A',
 'Not assigned',
 'Not assigned\n',
 'M9A',
 'Etobicoke',
 'Islington Avenue\n',
 'M1B',
 'Scarborough',
 'Rouge\n',
 'M1B',
 'Scarborough',
 'Malvern\n',
 'M2B',
 'Not assigned',
 'Not assigned\n',
 'M3B',
 'North York',
 'Don Mills North\n',
 'M4B',
 'East York',
 'Woodbine Gardens\n',
 'M4B',
 'East York',
 'Parkview Hill\n',
 'M5B',
 'Downtown Toronto',
 'Ryerson\n',
 'M5B',
 'Downtown Toronto',
 'Garden District\n',
 'M6B',
 'North York',
 'Glencairn\n',
 'M7B',
 'Not assigned',
 'Not assigned\n',
 'M8B',
 'Not assigned',
 'Not assigned\n',
 'M9B',
 'Etobicoke',
 'Cloverdale\n',

### 6. We shall pick the data one by one from the table in HTML and insert it into the dataframe

In [7]:
postcode = []
borough = []
Neighborhood = [] #lists defined
for i in range(0,len(info),3):
    if info[i+1] == 'Not assigned' :   
        pass #ignoring all the cells that have Not assigned in the Borough field
    else :
        postcode.append(info[i]) #creating the postcode list
        borough.append(info[i+1]) #creating the borough list
        if info[i+2] == 'Not assigned\n' : 
            Neighborhood.append(info[i+1]) #if Neighborhood is Not assigned then make it same as the borough
        else :
            Neighborhood.append(info[i+2].strip('\n')) #otherwise keep it as it is

In [8]:
print(postcode)
print("******************************************************************************************************************")
print(borough)
print("******************************************************************************************************************")
print(Neighborhood)

['M3A', 'M4A', 'M5A', 'M5A', 'M6A', 'M6A', 'M7A', 'M9A', 'M1B', 'M1B', 'M3B', 'M4B', 'M4B', 'M5B', 'M5B', 'M6B', 'M9B', 'M9B', 'M9B', 'M9B', 'M9B', 'M1C', 'M1C', 'M1C', 'M3C', 'M3C', 'M4C', 'M5C', 'M6C', 'M9C', 'M9C', 'M9C', 'M9C', 'M1E', 'M1E', 'M1E', 'M4E', 'M5E', 'M6E', 'M1G', 'M4G', 'M5G', 'M6G', 'M1H', 'M2H', 'M3H', 'M3H', 'M3H', 'M4H', 'M5H', 'M5H', 'M5H', 'M6H', 'M6H', 'M1J', 'M2J', 'M2J', 'M2J', 'M3J', 'M3J', 'M4J', 'M5J', 'M5J', 'M5J', 'M6J', 'M6J', 'M1K', 'M1K', 'M1K', 'M2K', 'M3K', 'M3K', 'M4K', 'M4K', 'M5K', 'M5K', 'M6K', 'M6K', 'M6K', 'M1L', 'M1L', 'M1L', 'M2L', 'M2L', 'M3L', 'M4L', 'M4L', 'M5L', 'M5L', 'M6L', 'M6L', 'M6L', 'M9L', 'M1M', 'M1M', 'M1M', 'M2M', 'M2M', 'M3M', 'M4M', 'M5M', 'M5M', 'M6M', 'M6M', 'M6M', 'M6M', 'M9M', 'M9M', 'M1N', 'M1N', 'M2N', 'M3N', 'M4N', 'M5N', 'M6N', 'M6N', 'M9N', 'M1P', 'M1P', 'M1P', 'M2P', 'M4P', 'M5P', 'M5P', 'M6P', 'M6P', 'M9P', 'M1R', 'M1R', 'M2R', 'M4R', 'M5R', 'M5R', 'M5R', 'M6R', 'M6R', 'M7R', 'M9R', 'M9R', 'M9R', 'M9R', 'M1S', 'M4S'

#### all our data looks good

### 7. Defining the dataframe

In [9]:
# define the dataframe columns
column_names = ['PostalCode','Borough','Neighborhood']

# instantiate the dataframe
neighborhoods = pd.DataFrame(columns=column_names)
neighborhoods

Empty DataFrame
Columns: [PostalCode, Borough, Neighborhood]
Index: []

### Checking the lengths of the lists so far(whether they are of same length)

In [10]:
print(len(postcode))
print(len(borough))
print(len(Neighborhood))

212
212
212


In [11]:
for i in range(0,len(postcode)-1):
    j = i+1
    while postcode[i]==postcode[j]:
        Neighborhood[i] = Neighborhood[i] + ',' + Neighborhood[j] 
        j+=1    
        

IndexError: list index out of range

In [12]:
def reverseWords(input): 
    inputWords = input.split(",")
    inputWords=inputWords[-1::-1]
    output = ','.join(inputWords)
    return output

In [13]:
reverseWords('Harbourfront,Regent Park')

'Regent Park,Harbourfront'

In [14]:
neighborhood = []
for i in Neighborhood:
    neighborhood.append(reverseWords(i))

### Let's Check the Neighborhood list now

In [15]:
for i in neighborhood:
    print(i)

Parkwoods
Victoria Village
Regent Park,Harbourfront
Regent Park
Lawrence Manor,Lawrence Heights
Lawrence Manor
Queen's Park
Islington Avenue
Malvern,Rouge
Malvern
Don Mills North
Parkview Hill,Woodbine Gardens
Parkview Hill
Garden District,Ryerson
Garden District
Glencairn
West Deane Park,Princess Gardens,Martin Grove,Islington,Cloverdale
West Deane Park,Princess Gardens,Martin Grove,Islington
West Deane Park,Princess Gardens,Martin Grove
West Deane Park,Princess Gardens
West Deane Park
Port Union,Rouge Hill,Highland Creek
Port Union,Rouge Hill
Port Union
Don Mills South,Flemingdon Park
Don Mills South
Woodbine Heights
St. James Town
Humewood-Cedarvale
Old Burnhamthorpe,Markland Wood,Eringate,Bloordale Gardens
Old Burnhamthorpe,Markland Wood,Eringate
Old Burnhamthorpe,Markland Wood
Old Burnhamthorpe
West Hill,Morningside,Guildwood
West Hill,Morningside
West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Wilson Hei

In [16]:
print(len(postcode))
print(len(borough))
print(len(neighborhood))

212
212
212


### The lengths are still 212

### 8. Insert the lists into a dataframe 

In [17]:
for a,b,c in zip(postcode,borough,neighborhood):    
    neighborhoods = neighborhoods.append({'PostalCode' : a,
                                          'Borough': b,
                                          'Neighborhood': c}, ignore_index=True)

In [18]:
neighborhoods

PostalCode           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M5A  Downtown Toronto   
4          M6A        North York   
5          M6A        North York   
6          M7A      Queen's Park   
7          M9A         Etobicoke   
8          M1B       Scarborough   
9          M1B       Scarborough   
10         M3B        North York   
11         M4B         East York   
12         M4B         East York   
13         M5B  Downtown Toronto   
14         M5B  Downtown Toronto   
15         M6B        North York   
16         M9B         Etobicoke   
17         M9B         Etobicoke   
18         M9B         Etobicoke   
19         M9B         Etobicoke   
20         M9B         Etobicoke   
21         M1C       Scarborough   
22         M1C       Scarborough   
23         M1C       Scarborough   
24         M3C        North York   
25         M3C        North York   
26         M4C         East York   
27         M5C  Downtown Toronto   
28         M6C              York   
29         M9C         Etobicoke   
..         ...               ...   
182        M1W       Scarborough   
183        M1W       Scarborough   
184        M4W  Downtown Toronto   
185        M5W  Downtown Toronto   
186        M8W         Etobicoke   
187        M8W         Etobicoke   
188        M9W         Etobicoke   
189        M1X       Scarborough   
190        M4X  Downtown Toronto   
191        M4X  Downtown Toronto   
192        M5X  Downtown Toronto   
193        M5X  Downtown Toronto   
194        M8X         Etobicoke   
195        M8X         Etobicoke   
196        M8X         Etobicoke   
197        M4Y  Downtown Toronto   
198        M7Y      East Toronto   
199        M8Y         Etobicoke   
200        M8Y         Etobicoke   
201        M8Y         Etobicoke   
202        M8Y         Etobicoke   
203        M8Y         Etobicoke   
204        M8Y         Etobicoke   
205        M8Y         Etobicoke   
206        M8Y         Etobicoke   
207        M8Z         Etobicoke   
208        M8Z         Etobicoke   
209        M8Z         Etobicoke   
210        M8Z         Etobicoke   
211        M8Z         Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                             Regent Park,Harbourfront  
3                                          Regent Park  
4                      Lawrence Manor,Lawrence Heights  
5                                       Lawrence Manor  
6                                         Queen's Park  
7                                     Islington Avenue  
8                                        Malvern,Rouge  
9                                              Malvern  
10                                     Don Mills North  
11                      Parkview Hill,Woodbine Gardens  
12                                       Parkview Hill  
13                             Garden District,Ryerson  
14                                     Garden District  
15                                           Glencairn  
16   West Deane Park,Princess Gardens,Martin Grove,...  
17   West Deane Park,Princess Gardens,Martin Grove,...  
18       West Deane Park,Princess Gardens,Martin Grove  
19                    West Deane Park,Princess Gardens  
20                                     West Deane Park  
21                Port Union,Rouge Hill,Highland Creek  
22                               Port Union,Rouge Hill  
23                                          Port Union  
24                     Don Mills South,Flemingdon Park  
25                                     Don Mills South  
26                                    Woodbine Heights  
27                                      St. James Town  
28                                  Humewood-Cedarvale  
29   Old Burnhamthorpe,Markland Wood,Eringate,Bloor...  
..  

### 9. Now if we filter out the duplicate columns and keep the first occurrences we can get the dataframe as asked in the question

In [19]:
neighborhoods.drop_duplicates(subset='PostalCode', keep='first', inplace=True)

In [20]:
neighborhoods = neighborhoods.reset_index()

In [21]:
neighborhoods = neighborhoods.drop('index',axis = 1)

In [22]:
neighborhoods

PostalCode           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A      Queen's Park   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   
12         M1C       Scarborough   
13         M3C        North York   
14         M4C         East York   
15         M5C  Downtown Toronto   
16         M6C              York   
17         M9C         Etobicoke   
18         M1E       Scarborough   
19         M4E      East Toronto   
20         M5E  Downtown Toronto   
21         M6E              York   
22         M1G       Scarborough   
23         M4G         East York   
24         M5G  Downtown Toronto   
25         M6G  Downtown Toronto   
26         M1H       Scarborough   
27         M2H        North York   
28         M3H        North York   
29         M4H         East York   
..         ...               ...   
73         M4R   Central Toronto   
74         M5R   Central Toronto   
75         M6R      West Toronto   
76         M7R       Mississauga   
77         M9R         Etobicoke   
78         M1S       Scarborough   
79         M4S   Central Toronto   
80         M5S  Downtown Toronto   
81         M6S      West Toronto   
82         M1T       Scarborough   
83         M4T   Central Toronto   
84         M5T  Downtown Toronto   
85         M1V       Scarborough   
86         M4V   Central Toronto   
87         M5V  Downtown Toronto   
88         M8V         Etobicoke   
89         M9V         Etobicoke   
90         M1W       Scarborough   
91         M4W  Downtown Toronto   
92         M5W  Downtown Toronto   
93         M8W         Etobicoke   
94         M9W         Etobicoke   
95         M1X       Scarborough   
96         M4X  Downtown Toronto   
97         M5X  Downtown Toronto   
98         M8X         Etobicoke   
99         M4Y  Downtown Toronto   
100        M7Y      East Toronto   
101        M8Y         Etobicoke   
102        M8Z         Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                             Regent Park,Harbourfront  
3                      Lawrence Manor,Lawrence Heights  
4                                         Queen's Park  
5                                     Islington Avenue  
6                                        Malvern,Rouge  
7                                      Don Mills North  
8                       Parkview Hill,Woodbine Gardens  
9                              Garden District,Ryerson  
10                                           Glencairn  
11   West Deane Park,Princess Gardens,Martin Grove,...  
12                Port Union,Rouge Hill,Highland Creek  
13                     Don Mills South,Flemingdon Park  
14                                    Woodbine Heights  
15                                      St. James Town  
16                                  Humewood-Cedarvale  
17   Old Burnhamthorpe,Markland Wood,Eringate,Bloor...  
18                     West Hill,Morningside,Guildwood  
19                                         The Beaches  
20                                         Berczy Park  
21                                 Caledonia-Fairbanks  
22                                              Woburn  
23                                             Leaside  
24                                  Central Bay Street  
25                                            Christie  
26                                           Cedarbrae  
27                                   Hillcrest Village  
28       Wilson Heights,Downsview North,Bathurst Manor  
29                                    Thorncliffe Park  
..  

Checking whether it looks like the one given in the question.

In [23]:
# define the dataframe columns
column_names = ['PostalCode','Borough','Neighborhood']

# instantiate the dataframe
data1 = pd.DataFrame(columns=column_names)
data1

list1 = ['M5G', 'M2H', 'M4B', 'M1J', 'M4G', 'M4M', 'M1R', 'M9V', 'M9L', 'M5V', 'M1B', 'M5A']
for post in list1 :
    data1 = data1.append(neighborhoods[neighborhoods['PostalCode']==post], ignore_index=True)
    
data1

PostalCode           Borough  \
0         M5G  Downtown Toronto   
1         M2H        North York   
2         M4B         East York   
3         M1J       Scarborough   
4         M4G         East York   
5         M4M      East Toronto   
6         M1R       Scarborough   
7         M9V         Etobicoke   
8         M9L        North York   
9         M5V  Downtown Toronto   
10        M1B       Scarborough   
11        M5A  Downtown Toronto   

                                         Neighborhood  
0                                  Central Bay Street  
1                                   Hillcrest Village  
2                      Parkview Hill,Woodbine Gardens  
3                                 Scarborough Village  
4                                             Leaside  
5                                     Studio District  
6                                    Wexford,Maryvale  
7   Thistletown,South Steeles,Silverstone,Mount Ol...  
8                                       Humber Summit  
9   South Niagara,Railway Lands,King and Spadina,H...  
10                                      Malvern,Rouge  
11                           Regent Park,Harbourfront

# Voila! We have the dataframe that was asked in the question

### 10. Checking the shape of the dataframe

In [24]:
neighborhoods.shape

(103, 3)

### Converting the dataframe into a CSV file so that we can use it in next part of the assignment

In [25]:
neighborhoods.to_csv('neighborhoods.csv')